tscat/__init__.py

```diff
@@ -46,61 +59,201 @@ def variable_attributes_as_dict(self):
+                return False
+        return True
+
+
+@typechecked
 class Event(_BackendBasedEntity):
-    _fixed_keys = ['start', 'stop', 'author', 'uuid']
+    _fixed_keys = ['start', 'stop', 'author', 'uuid', 'tags', 'products']
 
-    def __init__(self, start: dt.datetime, stop: dt.datetime, author: str, uuid=None, **kwargs):
+    def __init__(self, start: dt.datetime, stop: dt.datetime,
+                 author: str,
+                 uuid: str = None,
+                 tags: Iterable[str] = [],
+                 products: Iterable[str] = [],
+                 _insert: bool = True,
+                 **kwargs):
+        self._in_ctor = True
         super().__init__()
 
-        if start >= stop:
-            raise ValueError("start-datetime has to be younger than end-datetime")
-
         self.start = start
         self.stop = stop
         self.author = author
+        self.tags = list(tags)
+        self.products = list(products)
 
         if not uuid:
             self.uuid = str(uuid4())
         else:
-            UUID(uuid, version=4)  # check whether it is a valid uuid - throws ValueError if not
             self.uuid = uuid
 
-        _create_attributes(self, kwargs)
+        _verify_attribute_names(kwargs)
+        self.__dict__.update(kwargs)
+
+        if _insert:
+            self._backend_entity = backend().add_or_update_event({
+                'start': self.start,
+                'stop': self.stop,
+                'author': self.author,
+                'uuid': self.uuid,
+                'tags': self.tags,
+                'products': self.products,
+                'attributes': kwargs,
+            })
+
+        self._in_ctor = False
+
+    def __setattr__(self, key, value):
+        if key == 'uuid':
+            UUID(value, version=4)  # throws an exception if not valid
+        elif key == 'start' and hasattr(self, 'stop'):
+            if value >= self.stop:
+                raise ValueError("start-datetime has to be younger than end-datetime")
+        elif key == 'stop' and hasattr(self, 'start'):
+            if value <= self.start:
+                raise ValueError("stop-datetime has to be younger than end-datetime")
+        elif key in ['tags', 'products']:
+            if any(type(v) != str for v in value):
+                raise ValueError("a tag has to be a string")
+
+        super(Event, self).__setattr__(key, value)
 
     def __repr__(self):
         return self.representation('Event')
 
 
+@typechecked
 class Catalogue(_BackendBasedEntity):
-    _fixed_keys = ['name', 'author', 'predicate']
+    _fixed_keys = ['name', 'author', 'tags', 'predicate']
 
     def __init__(self, name: str, author: str,
+                 tags: Iterable[str] = [],
                  predicate: Predicate = None,
                  events: List[Event] = None,
+                 _insert: bool = True,
                  **kwargs):
-        super().__init__()
+        self._in_ctor = True
```

Comment:
Why this weird thing/hack?

Reply:
The `__setattr__`-method of the base-class is updating the backend-entity upon class-attribute-changes. `_in_ctor` disables this while we are in the constructor - we don't want the backend to be updated while being the constructor.

I'd be happy to have better solution.